pkg/storegateway/indexheader/binary_reader.go

```diff
@@ -533,84 +534,79 @@ func newFileBinaryReader(path string, postingOffsetsInMemSampling int, cfg Binar
 		return nil, errors.Wrap(err, "read symbols")
 	}
 
-	var lastKey []string
+	var lastLbl labels.Label
+	var lastSet bool
 	if r.indexVersion == index.FormatV1 {
 		// Earlier V1 formats don't have a sorted postings offset table, so
 		// load the whole offset table into memory.
 		r.postingsV1 = map[string]map[string]index.Range{}
 
 		var prevRng index.Range
-		if err := index.ReadOffsetTable(r.b, r.toc.PostingsOffsetTable, func(key []string, off uint64, _ int) error {
-			if len(key) != 2 {
-				return errors.Errorf("unexpected key length for posting table %d", len(key))
-			}
-
-			if lastKey != nil {
+		if err := readPostingsOffsetTable(r.b, r.toc.PostingsOffsetTable, func(lbl labels.Label, off uint64, _ int) error {
+			if lastSet {
 				prevRng.End = int64(off - crc32.Size)
-				r.postingsV1[lastKey[0]][lastKey[1]] = prevRng
+				r.postingsV1[lastLbl.Name][lastLbl.Value] = prevRng
 			}
 
-			if _, ok := r.postingsV1[key[0]]; !ok {
-				r.postingsV1[key[0]] = map[string]index.Range{}
-				r.postings[key[0]] = nil // Used to get a list of labelnames in places.
+			if _, ok := r.postingsV1[lbl.Name]; !ok {
+				r.postingsV1[lbl.Name] = map[string]index.Range{}
+				r.postings[lbl.Name] = nil // Used to get a list of labelnames in places.
 			}
 
-			lastKey = key
+			lastSet = true
+			lastLbl = lbl
 			prevRng = index.Range{Start: int64(off + postingLengthFieldSize)}
 			return nil
 		}); err != nil {
 			return nil, errors.Wrap(err, "read postings table")
 		}
-		if lastKey != nil {
+		if lastLbl.Name != "" {
```

Comment:
Shouldn't this be: `if lastSet` ? (Although it doesn't change the logic, unless someone writes postings for empty label name)